File: packages/python-sdk/kintsugi_sdk/_canonical.py
"""Canonicalization helpers shared by the sync + async clients."""

from __future__ import annotations

import json
from typing import Any


def serialize_body_for_sign(body: Any) -> str:
    """Match the TS SDK's serializeBodyForSign exactly.

    None / empty dict / empty string → "" (server-side hmac.guard uses the same rule).
    Everything else → JSON.stringify-equivalent (compact, separators with no spaces).
    """
    if body is None:
        return ""
    if isinstance(body, str):
        return body
    if isinstance(body, dict) and len(body) == 0:
        return ""
    return json.dumps(body, ensure_ascii=False, separators=(",", ":"))

File: packages/java-sdk/src/main/java/com/kintsugi/sdk/Canonical.java
package com.kintsugi.sdk;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.util.Map;

/**
 * Canonicalization rules shared with the TS / Python / Go SDKs.
 * <p>
 * canonical = METHOD + "\n" + path + "\n" + timestamp + "\n" + nonce + "\n" + body
 * <p>
 * body == "" when the payload is null / empty Map / empty String;
 * otherwise body is the compact-JSON encoding of the payload.
 * The server-side hmac.guard uses identical rules, so this MUST match byte-for-byte.
 */
final class Canonical {
    private static final ObjectMapper MAPPER;

    static {
        MAPPER = new ObjectMapper();
        // compact output (no INDENT_OUTPUT) — same as JSON.stringify default
        MAPPER.disable(SerializationFeature.INDENT_OUTPUT);
        // do not escape non-ASCII (like ensure_ascii=False in Python)
        // Jackson default already does NOT escape non-ASCII.
    }

    private Canonical() {}

    static String serializeBodyForSign(Object body) {
        if (body == null) return "";
        if (body instanceof String s) return s;
        if (body instanceof Map<?, ?> m && m.isEmpty()) return "";
        try {
            return MAPPER.writeValueAsString(body);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("kintsugi: marshal body: " + e.getMessage(), e);
        }
    }

    /** Test-only access; package-private. */
    static ObjectMapper mapper() {
        return MAPPER;
    }
}
